### PR TITLE
SpectrumInfo rollout: <ConvertToConstantL2.cpp to SetScalingPSD.cpp>

### LINKED ISSUE #18685: SpectrumInfo and DetectorInfo rollout: <ConvertToConstantL2.cpp to SetScalingPSD.cpp1>
This issue is part of the SpectrumInfo and DetectorInfo rollout, see #17743.

In addition, the following set requires moving from using `ComponentHelper::moveComponent()` to `DetectorInfo::setPosition` and `ComponentHelper::rotateComponent()` to `DetectorInfo::setRotation`

Algorithms covered in this issue are:

- Framework/Algorithms/src/ConvertToConstantL2.cpp
- Framework/API/test/MatrixWorkspaceTest.h
- Framework/DataHandling/src/LoadDetectorInfo.cpp
- Framework/DataHandling/src/LoadHelper.cpp
- Framework/DataHandling/src/LoadILLIndirect.cpp
- Framework/DataHandling/src/LoadILLIndirect2.cpp
- Framework/DataHandling/src/LoadIsawDetCal.cpp
- Framework/DataHandling/src/MoveInstrumentComponent.cpp
- Framework/DataHandling/src/RotateInstrumentComponent.cpp
- Framework/DataHandling/src/SetScalingPSD.cpp1

### PR DESCRIPTION
**Pull request by PranavBahuguna**
This PR is part of refactoring effort to use the new interface of the SpectrumInfo module.
See #17743 for details.

In addition, the following set has been changed from using `ComponentHelper::moveComponent` to `DetectorInfo::setPosition` and `ComponentHelper::rotateComponent` to `DetectorInfo::setRotation`

Files changed include:
- ConvertToConstantL2.cpp
- LoadDetectorInfo.cpp
- LoadHelper.cpp
- LoadILLIndirect.cpp
- LoadILLIndirect2.cpp

The other files mentioned in the issue were not altered as they proved impossible or impractical to change.

Performance tests were added for:
- LoadIsawDetCal.cpp

None of the altered algorithms show any significantly improved performance. Regarding `LoadIsawDetCal.cpp`, attempting to use the new `setPosition` method resulted in enormously decreased performance, so the original code has been left in with the changed code commented out.

Fixes #18685.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by SimonHeybrock — COMMENTED** [review]
Hi @PranavBahuguna!

Most changes look good, some comments below.

Regarding `LoadIsawDetCal`, I think I know what the problem is and we will fix that soon as part of further refactoring of the instrument-internal code. Can you revert the changes to `LoadIsawDetCal.cpp`, there is it not really a good way to do this now, so as you suggested we leave this file alone for now?
Please keep the performance test, it will be useful for us when we do instrument changes, so thanks for adding that!

File: Framework/Algorithms/src/ConvertToConstantL2.cpp [hunk 1]
```diff
@@ -92,24 +89,22 @@ void ConvertToConstantL2::exec() {
...
     // position - set all detector distance to constant l2
     double r, theta, phi;
-    V3D oldPos = spectrumInfo.position(i);
+    V3D oldPos = inputSpecInfo.position(i);
     oldPos.getSpherical(r, theta, phi);
     V3D newPos;
     newPos.spherical(m_l2, theta, phi);
-    ComponentHelper::moveComponent(spectrumInfo.detector(i), pmap, newPos,
-                                   ComponentHelper::Absolute);
+    outputDetInfo.setPosition(i, newPos);
```
Comment (OwenArnold) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: Looking at the implementation. I don't think that a `DetectorGroup` can have it's own parameters. https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/DetectorGroup.cpp I believe that all `DetectorGroups` are created dynamically based on mapping, such as within `MatrixWorkspace.cpp`

@martyngigg Can you add anything further to this?

Comment (martyngigg) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: The main interface to `ParameterMap` for users is `SetInstrumentParameter` or the `Move/Rotate` algorthms and they only work with the instrument object and a list of names/detector Ids so it will never get a `DetectorGroup` back. Thinking about it `DetectorGroup::getComponentID` should probably throw really as they are not supposed to be considered "real".

Comment (OwenArnold) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: Would it be sensible to (a) change `DetectorGroup::getComponentID` as per @martyngigg  suggestion. But (b) also to check that `DetectorGroups` are never allowed to be entered into a Parameter map, if that check doesn't already exist.

Comment (SimonHeybrock) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: @OwenArnold Yes that makes sense, but I fear it will cause some algorithms/workflows to fail. Maybe do it in a separate PR?

@PranavBahuguna Regarding this PR, I would suggest adding something like:

```cpp
if(!spectrumInfo.hasUniqueDetector(i))
  throw ...

const auto detIndex = spectrumInfo.spectrumDefinition(i)[0].first;
outputDetInfo.setPosition(detIndex, newPos);
```

Comment (OwenArnold) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: 👍

Comment (martyngigg) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: I think having `DetectorGroup::getComponentID` throw should catch all of those situations so that you won't need the check? I agree that a new PR for that is what we want.

File: Framework/Algorithms/src/ConvertToConstantL2.cpp [hunk 2]
```diff
@@ -104,7 +114,9 @@ void ConvertToConstantL2::exec() {
     oldPos.getSpherical(r, theta, phi);
     V3D newPos;
     newPos.spherical(m_l2, theta, phi);
-    outputDetInfo.setPosition(i, newPos);
+
+    const size_t detIndex = inputSpecInfo.detector(i).index();
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: Don't get the detector, that is unnecessary overhad, and `Detector` will eventually be removed. Prefer using `SpectrumInfo::spectrumDefinition`.

Comment (PranavBahuguna) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: How do I access `SpectrumInfo::spectrumDefinition` though? Is there something I need to import for it?

Comment (SimonHeybrock) [inline Framework/Algorithms/src/ConvertToConstantL2.cpp]: @PranavBahuguna It is in a branch that was merged into master last week. If your changes are based on an older version, you may need to merge in master first.

**Comment by SimonHeybrock** [comment]
Thanks :+1:, code looks good now. :shipit: if Jenkins gives his ok.
